src/Module/RenderConfirmationBookingsHandler.php

```diff
@@ -0,0 +1,279 @@
+            'table_heading'        => $this->__('Bookings'),
+            'service_column'       => $this->__('Service'),
+            'booking_start_column' => $this->__('Start'),
+            'booking_end_column'   => $this->__('End'),
+            'booking_tz_column'    => $this->__('Timezone'),
+            'booking_rows'         => $rows,
+        ]);
+    }
+
+    /**
+     * Renders a booking row.
+     *
+     * @since [*next-version*]
+     *
+     * @param BookingInterface $booking The booking instance.
+     *
+     * @return string|Stringable The rendered booking row.
+     */
+    protected function _renderBookingRow(BookingInterface $booking)
+    {
+        $timezone = $this->_getDisplayTimezone($booking);
+
+        // Get timestamps from booking
+        $startTs = $booking->getStart();
+        $endTs   = $booking->getEnd();
+
+        // Create date time helper instances
+        $startDt = Carbon::createFromTimestampUTC($startTs);
+        $endDt   = Carbon::createFromTimestampUTC($endTs);
+
+        // Shift to client timezone, if available
+        if ($timezone !== null) {
+            $startDt->setTimezone($timezone);
+            $endDt->setTimezone($timezone);
+        }
+
+        // Format times to strings
+        $startStr = $startDt->format($this->bookingFormat);
+        $endStr   = $endDt->format($this->bookingFormat);
+
+        return $this->bookingRowTemplate->render([
+            'service_name' => $this->_getServiceName($booking),
+            'start_text'   => $startStr,
+            'end_text'     => $endStr,
+            'timezone'     => $timezone->getName(),
+        ]);
+    }
+
+    /**
+     * Retrieves the name of the service for a booking.
+     *
+     * @since [*next-version*]
+     *
+     * @param BookingInterface $booking The booking instance.
+     *
+     * @return string|Stringable The name of the service.
+     *
+     * @throws RuntimeException If failed to retrieve the service name or the service ID from the booking.
+     */
+    protected function _getServiceName(BookingInterface $booking)
+    {
+        try {
+            $container = $this->_normalizeContainer($booking);
+            $serviceId = $this->_containerGet($container, 'service_id');
+        } catch (Exception $exception) {
+            /*
+             * Catches:
+             * - Invalid argument exception if the booking is not a container
+             * - Container exception if an error occurred while reading
+             * - Container not found exception if the service ID was not found
+             */
+            throw $this->_createRuntimeException(
+                $this->__('Failed to get the ID of the service for a booking'), null, $exception
+            );
+        }
+
+        // Alias expression builder
+        $b = $this->exprBuilder;
+        // EDD Bookings' services select RM only supports AND top-level expressions
+        $services = $this->servicesSelectRm->select($b->and(
```

Comment:
Another case which should be optimized - see [card](https://github.com/orgs/RebelCode/projects/2#card-10541113).